crates/wysiwyg/src/node_joiner.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+
+use crate::dom::nodes::DomNode;
+use crate::dom::{Dom, DomHandle, DomLocation, MultipleNodesRange, Range};
+use crate::UnicodeString;
+
+/// Handles joining together nodes after an edit event.
+///
+/// For example, if your selection starts in a bold tag, leaves that tag,
+/// and then ends up in another bold tag, the final result should be a
+/// single bold tag containing all your text.
+pub(crate) struct NodeJoiner<S>
+where
+    S: UnicodeString,
+{
+    first_type: Option<String>,
+    last_type: Option<String>,
```

Comment:
For this I would just use `ContainerNodeKind` so we can merge `<b>` and `<strong>`, but it depends on whether we want to do it or not. This approach would also need extra checks for `Formatting` kind, i.e.

Reply:
Yeah, that sounds good - I'll look next week